SPDX-License-Identifier: GPL-2.0-only

diff --git a/drivers/input/misc/aw8697_haptic/aw8697.c b/drivers/input/misc/aw8697_haptic/aw8697.c
index 16d7f02fc86e..6f9e6b126e1a 100644
--- a/drivers/input/misc/aw8697_haptic/aw8697.c
+++ b/drivers/input/misc/aw8697_haptic/aw8697.c
@@ -4252,8 +4252,8 @@ static int aw8697_parse_dt_common(struct device *dev,
 			printk("%s Read qcom,wf-pattern property failed !\n",
 			       __func__);
 		}
-		printk("%s %d  effect->pattern_length=%d  effect->pattern=%d \n", __func__, __LINE__,
-			effect->pattern_length, (int)effect->pattern);
+		printk("%s %d  effect->pattern_length=%d  effect->pattern=%p \n", __func__, __LINE__,
+			effect->pattern_length, effect->pattern);
 
 		effect->play_rate_us = config->play_rate_us;
 		rc = of_property_read_u32(child_node, "qcom,wf-play-rate-us",
