diff --git a/drivers/power/supply/ti/bq2597x_charger_munch.c b/drivers/power/supply/ti/bq2597x_charger_munch.c
index 1b5ac6c3e88b..a7b0319c524d 100644
--- a/drivers/power/supply/ti/bq2597x_charger_munch.c
+++ b/drivers/power/supply/ti/bq2597x_charger_munch.c
@@ -2523,7 +2523,7 @@ static int bq2597x_get_dev_role(struct i2c_client *client)
 
 	dev_info(&client->dev, "%s: matched to %s\n", __func__, of_id->compatible);
 
-	return (int)of_id->data;
+	return (int)(long)of_id->data;
 }
 
 static int bq2597x_charger_probe(struct i2c_client *client,
