SPDX-License-Identifier: GPL-2.0-only

diff --git a/drivers/power/supply/ti/ln8000_charger_munch.c b/drivers/power/supply/ti/ln8000_charger_munch.c
index 3dc93021edb2..9ea099c6f201 100644
--- a/drivers/power/supply/ti/ln8000_charger_munch.c
+++ b/drivers/power/supply/ti/ln8000_charger_munch.c
@@ -1652,7 +1652,7 @@ static int ln8000_get_dev_role(struct i2c_client *client)
 
 	dev_info(&client->dev, "%s: matched to %s\n", __func__, of_id->compatible);
 
-	return (int)of_id->data;
+	return (int)(long)of_id->data;
 }
 
 static int ln8000_parse_dt(struct ln8000_info *info)
